src/mlpack/core/tree/cover_tree/cover_tree_impl.hpp

```diff
@@ -547,6 +548,74 @@ CoverTree<MetricType, StatisticType, MatType, RootPointPolicy>::CoverTree(
   }
 }
 
+// Copy Assignment.
+template<
+    typename MetricType,
+    typename StatisticType,
+    typename MatType,
+    typename RootPointPolicy
+>
+CoverTree<MetricType, StatisticType, MatType, RootPointPolicy>&
+CoverTree<MetricType, StatisticType, MatType, RootPointPolicy>::
+operator=(const CoverTree& other)
+{
+  if (this == &other)
+    return *this;
+
+  // Freeing memory that will not be used anymore.
+  delete dataset;
+  delete metric;
+  for (size_t i = 0; i < children.size(); ++i)
+  {
+    delete children[i];
+  }
+  children.clear();
+
+  dataset = ((other.parent == NULL && other.localDataset) ?
+      new MatType(*other.dataset) : other.dataset);
+  point = other.point;
+  scale = other.scale;
+  base = other.base;
+  stat = other.stat;
+  numDescendants = other.numDescendants;
+  parent = other.parent;
+  parentDistance = other.parentDistance;
+  furthestDescendantDistance = other.furthestDescendantDistance;
+  localMetric = false;
```

Comment:
I am not sure whether `localMetric` will always be false. And also if it's not always false then we need to make `metric` data member on `heap`. @rcurtin let me know if I am wrong as I am not sure of it. 
Also if I am correct then the same thing needs to be done in the `copy constructor (Line 520/521)` of `cover tree`. i.e if `localMetric` is true then we need to assign `metric` term a `new` memory on heap similar to what is done in the `constructor(line 253, 61)` to make a deep copy successful.

Reply:
@walragatver you're right, I think that this is a bug in the existing copy constructor.  Personally I think the way we should do it is that if `other.localMetric` is `true`, then we should set `localMetric` to `true` and set `metric` to a newly-allocated metric (and call the copy constructor of the metric).  @sriramsk1999 do you think you could make this change also?